### PR TITLE
CSV Concatenator

### LINKED ISSUE #775: CSV concatenator demo
Working on extractor implementation.

CSV Concatenator [Max]
When user uploads a CSV into a dataset, automatically update a single concatenated file that is the combination of all the CSV files in the dataset. Use Pandas to merge columns, allow user to provide column mappings somehow to help the extractor out. Could be more specific and require a “raw files” folder as the source or something along those lines.
Many projects have manual or periodically collected tabular data (e.g. sensor data) that someone has to manually combine into one large file. Rather than having a custom extractor to do this combination BEFORE upload as some projects do, have this happen seamlessly as soon as user uploads a new piece.

Vocabulary Standardizer [Max]
Somewhere we have a lookup table that has a list of synonyms for terms, maybe a lookup file or metadata entry? Then any text data that is uploaded will check contents and generate an equivalent file with the terms replaced with preferred synonym term.
Low-effort data cleaning that would let people upload data from various sources and have them all report data in a standard way without too much custom code.

### PR DESCRIPTION
**Pull request by max-zilla**
This adds an extractor that will try to concatenate CSV/TSV/XLSX files when uploaded to Clowder.

To test:
```
export CLOWDER_VERSION=2
python concatenate.py
```
(there's also a Dockerfile)

Then upload two CSVs to a dataset and see that concatenated.csv is created. If you already have 2+ CSVs in the dataset and you upload a new one, all will be merged into the output. Uploading additional CSVs will update the concatenated file as it goes.

File types are currently separated, meaning CSVs will get their own merge separate from Excel etc. This would be easy to change but not sure if there's a good use case for that?
image

Uses pyclowder files.delete so requires this version of pyclowder: https://github.com/clowder-framework/pyclowder/pull/92

### CODE REVIEW

File: csv-concatenator/extractor_info.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "@context": "https://clowderframework.org/contexts/extractors.jsonld",
+  "name": "ncsa.csv.concatenator",
+  "version": "1.0.0",
+  "description": "Automatically concatenate tabular data into an ongoing merged file.",
+  "author": "Max Burnette <mburnet2@illinois.edu>",
+  "contexts": [],
+  "repository": [],
```
Comment (longshuicy) [inline csv-concatenator/extractor_info.json]: Do you mind add this repository to the repository section?

File: csv-concatenator/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+pyclowder==3.0.6
```
Comment (longshuicy) [inline csv-concatenator/requirements.txt]: the latest pyclowder is 3.0.7 now. Maybe say >=3.0.7?

File: csv-concatenator/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.8
+
+ENV MAIN_SCRIPT="concatenator.py" \
+    RABBITMQ_QUEUE=""
+
+COPY requirements.txt ./
+RUN pip install -r requirements.txt
+
+COPY concatenator.py extractor_info.json ./
+CMD python concatenator.py
```
Comment (longshuicy) [inline csv-concatenator/Dockerfile:10]: Do you need to set the CLOWDER_VERSION as environment variable?

**Review by longshuicy — COMMENTED** [review]
Test works. But I haven't tested the built image in docker

**Review by ddey2 — APPROVED** [review]
I tested in pycharm. I had to install pandas and set Clowder_version. It works well. 

Approving this. I guess we can merge once you address the other comments.

File: csv-concatenator/Dockerfile (hunk 3 above)
Comment (max-zilla) [reply csv-concatenator/Dockerfile:10]: this extractor should work with both Clowder 1 and 2, so i don't want to hardcode a specific version here

File: csv-concatenator/requirements.txt (hunk 2 above)
Comment (max-zilla) [reply csv-concatenator/requirements.txt]: updated

File: csv-concatenator/extractor_info.json (hunk 1 above)
Comment (max-zilla) [reply csv-concatenator/extractor_info.json]: updated

**Pull request merged**